File: a2/tests/throughput.py
import requests
from concurrent.futures import ThreadPoolExecutor, as_completed
import time

BASE_URL = "http://127.0.0.1:8000"
NUM_REQUESTS = 2000
CONCURRENT_THREADS = 100

def post_request(index):
    """Send a POST request to create a user."""
    url = f"{BASE_URL}/user"
    data = {
        "command": "create",
        "id": index,
        "username": f"User{index}",
        "email": f"user{index}@example.com",
        "password": "password"  # Sending plain password
    }
    start_time = time.time()
    try:
        response = requests.post(url, json=data)
        end_time = time.time()
        if response.status_code == 200:
            return index, True, end_time - start_time
    except Exception as e:
        end_time = time.time()
    return index, False, end_time - start_time

def get_request(index):
    """Send a GET request to retrieve a user."""
    url = f"{BASE_URL}/user/{index}"
    start_time = time.time()
    try:
        response = requests.get(url)
        end_time = time.time()
        if response.status_code == 200:
            return True, end_time - start_time
    except Exception as e:
        end_time = time.time()
    return False, end_time - start_time

def perform_post_requests():
    """Perform POST requests using concurrent threads."""
    with ThreadPoolExecutor(max_workers=CONCURRENT_THREADS) as executor:
        futures = [executor.submit(post_request, i) for i in range(NUM_REQUESTS)]
        results = [future.result() for future in as_completed(futures)]
    return results

def perform_get_requests(successful_indices):
    """Perform GET requests for successfully created users."""
    with ThreadPoolExecutor(max_workers=CONCURRENT_THREADS) as executor:
        futures = [executor.submit(get_request, index) for index, success, _ in successful_indices if success]
        results = [future.result() for future in as_completed(futures)]
    return results

if __name__ == "__main__":
    total_start_time = time.time()

    # Perform POST requests and measure time
    post_start_time = time.time()
    post_results = perform_post_requests()
    post_end_time = time.time()
    print(f"Completed POST requests in {post_end_time - post_start_time:.2f} seconds.")

    # Extract successful POST request indices for corresponding GET requests
    successful_posts = [result for result in post_results if result[1]]

    # Perform GET requests and measure time
    get_start_time = time.time()
    get_results = perform_get_requests(successful_posts)
    get_end_time = time.time()
    print(f"Completed GET requests in {get_end_time - get_start_time:.2f} seconds.")

    total_end_time = time.time()
    print(f"Total operation time: {total_end_time - total_start_time:.2f} seconds.")

File: a2/tests/Throughput.java
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.http.HttpRequest.BodyPublishers;
import java.net.http.HttpResponse.BodyHandlers;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.time.Duration;

public class Throughput {

    private static final String BASE_URL = "http://127.0.0.1:8000";
    private static final int NUM_REQUESTS = 2000;
    private static final int CONCURRENT_THREADS = 100;
    private static final HttpClient client = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(10))
            .build();

    public static void main(String[] args) throws Exception {
        long totalStartTime = System.currentTimeMillis();

        // Perform POST requests
        long postStartTime = System.currentTimeMillis();
        List<CompletableFuture<Boolean>> postFutures = performPostRequests();
        CompletableFuture.allOf(postFutures.toArray(new CompletableFuture[0])).join();
        long postEndTime = System.currentTimeMillis();
        System.out.println("Completed POST requests in " + (postEndTime - postStartTime) + " milliseconds.");

        // Perform GET requests
        long getStartTime = System.currentTimeMillis();
        List<CompletableFuture<Boolean>> getFutures = performGetRequests();
        CompletableFuture.allOf(getFutures.toArray(new CompletableFuture[0])).join();
        long getEndTime = System.currentTimeMillis();
        System.out.println("Completed GET requests in " + (getEndTime - getStartTime) + " milliseconds.");

        long totalEndTime = System.currentTimeMillis();
        System.out.println("Total operation time: " + (totalEndTime - totalStartTime) + " milliseconds.");
    }

    private static List<CompletableFuture<Boolean>> performPostRequests() {
        ExecutorService executor = Executors.newFixedThreadPool(CONCURRENT_THREADS);
        List<CompletableFuture<Boolean>> futures = new ArrayList<>();
        for (int i = 0; i < NUM_REQUESTS; i++) {
            int index = i;
            CompletableFuture<Boolean> future = CompletableFuture.supplyAsync(() -> {
                try {
                    String json = "{\"command\": \"create\", \"id\": " + index + ", \"username\": \"User" + index + "\", \"email\": \"user" + index + "@example.com\", \"password\": \"password\"}";
                    HttpRequest request = HttpRequest.newBuilder()
                            .uri(URI.create(BASE_URL + "/user"))
                            .header("Content-Type", "application/json")
                            .POST(BodyPublishers.ofString(json))
                            .build();
                    HttpResponse<String> response = client.send(request, BodyHandlers.ofString());
                    return response.statusCode() == 200;
                } catch (Exception e) {
                    return false;
                }
            }, executor);
            futures.add(future);
        }
        return futures;
    }

    private static List<CompletableFuture<Boolean>> performGetRequests() {
        ExecutorService executor = Executors.newFixedThreadPool(CONCURRENT_THREADS);
        List<CompletableFuture<Boolean>> futures = new ArrayList<>();
        for (int i = 0; i < NUM_REQUESTS; i++) {
            int index = i;
            CompletableFuture<Boolean> future = CompletableFuture.supplyAsync(() -> {
                try {
                    HttpRequest request = HttpRequest.newBuilder()
                            .uri(URI.create(BASE_URL + "/user/" + index))
                            .build();
                    HttpResponse<String> response = client.send(request, BodyHandlers.ofString());
                    return response.statusCode() == 200;
                } catch (Exception e) {
                    return false;
                }
            }, executor);
            futures.add(future);
        }
        return futures;
    }
}

